huaweicloud/services/dws/resource_huaweicloud_dws_disaster_recovery.go

```diff
@@ -0,0 +1,566 @@
+	if err != nil {
+		return nil, err
+	}
+	return respBody, nil
+}
+
+func parsediasaterRecoveryError(err error) error {
+	var errCode400 golangsdk.ErrDefault400
+	if errors.As(err, &errCode400) {
+		var apiError interface{}
+		if jsonErr := json.Unmarshal(errCode400.Body, &apiError); jsonErr != nil {
+			return err
+		}
+		errorCode, errorCodeErr := jmespath.Search("error_code", apiError)
+		if errorCodeErr != nil {
+			return err
+		}
+
+		if errorCode == "DWS.10101" {
+			return golangsdk.ErrDefault404(errCode400)
+		}
+	}
+
+	return err
+}
+
+func resourceDwsDisasterRecoveryUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	var (
+		cfg     = meta.(*config.Config)
+		region  = cfg.GetRegion(d)
+		product = "dws"
+	)
+
+	client, err := cfg.NewServiceClient(product, region)
+	if err != nil {
+		return diag.Errorf("error creating DWS client: %s", err)
+	}
+
+	if d.HasChange("dr_sync_period") {
+		_, err = updateDisasterRecovery(client, d)
+		if err != nil {
+			return diag.Errorf("error updating DWS disaster recovery: %s", err)
+		}
+	}
+
+	if d.HasChange("action") {
+		action := d.Get("action").(string)
+		err := doAction(ctx, action, client, d)
+		if err != nil {
+			return err
+		}
+	}
+
+	return resourceDwsDisasterRecoveryRead(ctx, d, meta)
+}
+
+func doAction(ctx context.Context, action string, client *golangsdk.ServiceClient, d *schema.ResourceData) diag.Diagnostics {
+	switch action {
+	case "start":
+		err := doActionDisasterRecovery(client, d, action)
+		if err != nil {
+			return diag.Errorf("error starting DWS disaster recovery: %s", err)
+		}
+		// When the disaster recovery successfully started, the status is running.
+		err = waitingForActionDisasterRecvery(ctx, d, client, "running", d.Timeout(schema.TimeoutUpdate))
+		if err != nil {
+			return diag.Errorf("error waiting for the start of DWS disaster recovery (%s) to complete: %s", d.Id(), err)
+		}
+	case "pause":
+		err := doActionDisasterRecovery(client, d, action)
+		if err != nil {
+			return diag.Errorf("error stopping DWS disaster recovery: %s", err)
+		}
+		// When the disaster recovery successfully paused, the status is stopped.
+		err = waitingForActionDisasterRecvery(ctx, d, client, "stopped", d.Timeout(schema.TimeoutUpdate))
+		if err != nil {
+			return diag.Errorf("error waiting for the stop of DWS disaster recovery (%s) to complete: %s", d.Id(), err)
+		}
+	case "switchover":
+		err := doActionDisasterRecovery(client, d, action)
+		if err != nil {
+			return diag.Errorf("error switchovering DWS disaster recovery: %s", err)
+		}
+		// When the disaster recovery successfully swtichovered, the status is running.
+		err = waitingForActionDisasterRecvery(ctx, d, client, "running", d.Timeout(schema.TimeoutUpdate))
+		if err != nil {
+			return diag.Errorf("error waiting for the swtichover of DWS disaster recovery (%s) to complete: %s", d.Id(), err)
+		}
+	default:
+		return diag.Errorf("cannot marshal action: " + action)
```

Comment:
what does `marshal` mean in this context?

Reply:
done